File: .agents/skills/k2-design-system/scripts/test_validate_design_system.py
#!/usr/bin/env python3
"""Focused tests for the K2 design-system validator."""

from __future__ import annotations

import copy
import importlib.util
import json
import tempfile
import unittest
from pathlib import Path


SCRIPT_PATH = Path(__file__).with_name("validate-design-system.py")
SPEC = importlib.util.spec_from_file_location("k2_design_system_validator", SCRIPT_PATH)
if SPEC is None or SPEC.loader is None:
    raise RuntimeError(f"Unable to load validator from {SCRIPT_PATH}")
VALIDATOR = importlib.util.module_from_spec(SPEC)
SPEC.loader.exec_module(VALIDATOR)


class K2DesignSystemValidatorTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls) -> None:
        cls.root = SCRIPT_PATH.resolve().parents[1]
        cls.tokens = VALIDATOR.load_tokens(cls.root)

    def test_current_skill_is_valid(self) -> None:
        self.assertEqual(VALIDATOR.validate(self.root), [])

    def test_current_skill_uses_k2_identity(self) -> None:
        skill = (self.root / "SKILL.md").read_text(encoding="utf-8")
        interface = (self.root / "agents" / "openai.yaml").read_text(encoding="utf-8")
        self.assertIn("name: k2-design-system", skill)
        self.assertIn('display_name: "K2 Design System"', interface)
        self.assertIn("$k2-design-system", interface)

    def test_canonical_palette_values_cannot_drift(self) -> None:
        tokens = copy.deepcopy(self.tokens)
        tokens["color"]["spark-blue"]["$value"] = "#000000"
        errors: list[str] = []
        VALIDATOR.validate_tokens(tokens, errors)
        self.assertIn("color.spark-blue must equal #1CB0F6", errors)

    def test_action_roles_cannot_drift(self) -> None:
        tokens = copy.deepcopy(self.tokens)
        tokens["themes"]["light"]["action"]["primary"] = "#58CC02"
        tokens["themes"]["light"]["action"]["secondaryForeground"] = "#1CB0F6"
        tokens["themes"]["light"]["action"]["disabledBackground"] = "#AFAFAF"
        tokens["themes"]["dark"]["action"]["success"] = "#49C0F8"
        tokens["themes"]["dark"]["action"]["secondaryForeground"] = "#F0F7F2"
        errors: list[str] = []
        VALIDATOR.validate_tokens(tokens, errors)
        self.assertIn(
            "themes.light.action.primary must equal #1CB0F6",
            errors,
        )
        self.assertIn(
            "themes.light.action.secondaryForeground must equal #4B4B4B",
            errors,
        )
        self.assertIn(
            "themes.light.action.disabledBackground must equal #D9D9D9",
            errors,
        )
        self.assertIn(
            "themes.dark.action.success must equal #72D72B",
            errors,
        )
        self.assertIn(
            "themes.dark.action.secondaryForeground must equal #4B4B4B",
            errors,
        )

    def test_token_type_is_required(self) -> None:
        tokens = copy.deepcopy(self.tokens)
        del tokens["color"]["eager-green"]["$type"]
        errors: list[str] = []
        VALIDATOR.validate_tokens(tokens, errors)
        self.assertIn("color.eager-green must use type color", errors)

    def test_token_description_is_required(self) -> None:
        tokens = copy.deepcopy(self.tokens)
        tokens["font"]["feather"]["$description"] = ""
        errors: list[str] = []
        VALIDATOR.validate_tokens(tokens, errors)
        self.assertIn("font.feather must have a description", errors)

    def test_button_geometry_is_required(self) -> None:
        tokens = copy.deepcopy(self.tokens)
        tokens["$extensions"]["com.vocora.design-system"]["buttonRadius"] = "8px"
        errors: list[str] = []
        VALIDATOR.validate_tokens(tokens, errors)
        self.assertIn("The canonical button radius must be 13px", errors)

    def test_light_and_dark_theme_keys_require_parity(self) -> None:
        tokens = copy.deepcopy(self.tokens)
        del tokens["themes"]["dark"]["action"]["error"]
        errors: list[str] = []
        VALIDATOR.validate_tokens(tokens, errors)
        self.assertIn("Theme key parity failed for group 'action'", errors)

    def test_css_variables_match_the_canonical_palette(self) -> None:
        variables = VALIDATOR.extract_css_variables(
            (self.root / "references" / "variables.scss").read_text(
                encoding="utf-8"
            )
        )
        for name, value in VALIDATOR.CANONICAL_COLORS.items():
            self.assertEqual(variables[f"--color-{name}"].lower(), value.lower())

    def test_frontend_contract_defines_the_complete_decision_hierarchy(self) -> None:
        design = (self.root / "references" / "DESIGN.md").read_text(
            encoding="utf-8"
        )
        hierarchy = (
            "Existing Vocora shared primitive?",
            "Standard interactive primitive in Angular Material?",
            "Required behavior available in Angular CDK?",
            "Layout, spacing, display, or semantic utility in Bootstrap?",
            "Existing shared style or semantic token?",
            "Otherwise",
        )
        positions = [design.index(item) for item in hierarchy]
        self.assertEqual(positions, sorted(positions))
        self.assertIn("## Touch-to-refactor", design)
        self.assertIn("## Custom CSS last", design)
        self.assertIn("## Specificity and `!important`", design)

    def test_frontend_architecture_rejects_non_owner_theme_definitions(self) -> None:
        with tempfile.TemporaryDirectory() as directory:
            repo = Path(directory)
            self._write_baseline(repo)
            feature = repo / "ui/src/app/features/example/example.component.scss"
            feature.parent.mkdir(parents=True)
            feature.write_text(
                ":host { --bs-primary: red; --mat-sys-primary: blue; "
                "--color-spark-blue: green; }",
                encoding="utf-8",
            )
            errors: list[str] = []

            VALIDATOR.validate_frontend_architecture(repo, errors)

            self.assertIn(
                "Bootstrap semantic variables must be defined only in "
                "ui/src/styles/_bootstrap-theme.scss: "
                "ui/src/app/features/example/example.component.scss defines --bs-primary",
                errors,
            )
            self.assertIn(
                "Material system variables must be defined only in "
                "ui/src/styles/_angular-material-theme.scss: "
                "ui/src/app/features/example/example.component.scss defines --mat-sys-primary",
                errors,
            )
            self.assertIn(
                "Vocora foundation colors must be defined only in "
                "ui/src/styles/_vocora-design-system.scss: "
                "ui/src/app/features/example/example.component.scss defines --color-spark-blue",
                errors,
            )

    def test_frontend_architecture_rejects_unbaselined_specificity_debt(self) -> None:
        with tempfile.TemporaryDirectory() as directory:
            repo = Path(directory)
            self._write_baseline(repo)
            feature = repo / "ui/src/app/features/example/example.component.scss"
            feature.parent.mkdir(parents=True)
            feature.write_text(
                ".mat-mdc-example { color: red !important; }",
                encoding="utf-8",
            )
            errors: list[str] = []

            VALIDATOR.validate_frontend_architecture(repo, errors)

            self.assertIn(
                "Untracked !important debt in "
                "ui/src/app/features/example/example.component.scss: expected 0, found 1",
                errors,
            )
            self.assertIn(
                "Untracked feature Material-internal selector debt in "
                "ui/src/app/features/example/example.component.scss: expected 0, found 1",
                errors,
            )

    @staticmethod
    def _write_baseline(repo: Path) -> None:
        baseline = (
            repo
            / ".agents/skills/k2-design-system/references/legacy-style-baseline.json"
        )
        baseline.parent.mkdir(parents=True)
        baseline.write_text(
            json.dumps(
                {
                    "schema_version": 1,
                    "important_declaration_counts": {},
                    "feature_material_internal_selector_counts": {},
                }
            ),
            encoding="utf-8",
        )


if __name__ == "__main__":
    unittest.main()

File: .agents/skills/k2-design-system/scripts/test-validate-design-system.mjs
#!/usr/bin/env node

import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const skillRoot = path.resolve(scriptDir, "..");

function runValidator(root) {
  const result = spawnSync(
    "python3",
    [path.join(root, "scripts", "validate-design-system.py")],
    {
      encoding: "utf8",
    },
  );
  return { status: result.status, output: `${result.stdout}${result.stderr}` };
}

function withFixture(mutator, expectation) {
  const directory = fs.mkdtempSync(path.join(os.tmpdir(), "k2-design-system-"));
  const fixtureRoot = path.join(directory, "k2-design-system");
  try {
    fs.cpSync(skillRoot, fixtureRoot, { recursive: true });
    mutator(fixtureRoot);
    expectation(runValidator(fixtureRoot));
  } finally {
    fs.rmSync(directory, { recursive: true, force: true });
  }
}

function mutateTokens(root, mutator) {
  const tokenPath = path.join(root, "references", "tokens.json");
  const tokens = JSON.parse(fs.readFileSync(tokenPath, "utf8"));
  mutator(tokens);
  fs.writeFileSync(tokenPath, `${JSON.stringify(tokens, null, 2)}\n`);
}

const current = runValidator(skillRoot);
assert.equal(current.status, 0, current.output);

const skill = fs.readFileSync(path.join(skillRoot, "SKILL.md"), "utf8");
const interfaceSource = fs.readFileSync(
  path.join(skillRoot, "agents", "openai.yaml"),
  "utf8",
);
assert.match(skill, /name: k2-design-system/u);
assert.match(interfaceSource, /display_name: "K2 Design System"/u);
assert.match(interfaceSource, /\$k2-design-system/u);
assert.match(skill, /ui\/tests\/test-design-system-architecture\.mjs/u);

withFixture(
  (root) =>
    mutateTokens(root, (tokens) => {
      tokens.color["spark-blue"].$value = "#000000";
    }),
  ({ status, output }) => {
    assert.notEqual(status, 0);
    assert.match(output, /color\.spark-blue must equal #1CB0F6/u);
  },
);

withFixture(
  (root) => {
    const materialPath = path.join(root, "references", "material-theme.scss");
    fs.writeFileSync(
      materialPath,
      fs
        .readFileSync(materialPath, "utf8")
        .replace(
          "--mat-sys-inverse-primary: var(--vocora-inverse-primary);",
          "--mat-sys-inverse-primary: var(--vocora-primary);",
        ),
    );
  },
  ({ output }) =>
    assert.match(
      output,
      /material-theme\.scss must map --mat-sys-inverse-primary to --vocora-inverse-primary/u,
    ),
);

withFixture(
  (root) =>
    mutateTokens(root, (tokens) => {
      delete tokens.themes.dark.border.default;
      tokens.themes.light.semantic.secondary = tokens.themes.light.text.secondary;
    }),
  ({ output }) => {
    assert.match(output, /Theme key parity failed for group 'border'/u);
    assert.match(
      output,
      /themes\.light\.semantic\.secondary must equal #000437/u,
    );
  },
);

withFixture(
  (root) => {
    const skillPath = path.join(root, "SKILL.md");
    fs.writeFileSync(
      skillPath,
      fs
        .readFileSync(skillPath, "utf8")
        .replace("are removed from the tab order", "leave the tab order"),
    );
  },
  ({ status, output }) => {
    assert.notEqual(status, 0);
    assert.match(
      output,
      /SKILL\.md must say disabled Voco links are removed from the tab order/u,
    );
  },
);

withFixture(
  (root) => {
    const designPath = path.join(root, "references", "DESIGN.md");
    fs.writeFileSync(
      designPath,
      fs
        .readFileSync(designPath, "utf8")
        .replace("is removed from the tab order", "leaves the tab order"),
    );
  },
  ({ status, output }) => {
    assert.notEqual(status, 0);
    assert.match(
      output,
      /DESIGN\.md must say disabled Voco links are removed from the tab order/u,
    );
  },
);

withFixture(
  (root) =>
    mutateTokens(root, (tokens) => {
      tokens.themes.light.action.primary = "#58CC02";
      tokens.themes.light.action.secondaryForeground = "#1CB0F6";
      tokens.themes.light.action.disabledBackground = "#AFAFAF";
      tokens.themes.dark.action.success = "#49C0F8";
      tokens.themes.dark.action.secondaryForeground = "#F0F7F2";
    }),
  ({ output }) => {
    for (const expected of [
      "themes.light.action.primary must equal #1CB0F6",
      "themes.light.action.secondaryForeground must equal #4B4B4B",
      "themes.light.action.disabledBackground must equal #D9D9D9",
      "themes.dark.action.success must equal #72D72B",
      "themes.dark.action.secondaryForeground must equal #4B4B4B",
    ])
      assert.match(
        output,
        new RegExp(expected.replace(/[.*+?^${}()|[\]\\]/gu, "\\$&"), "u"),
      );
  },
);

withFixture(
  (root) =>
    mutateTokens(root, (tokens) => {
      delete tokens.color["eager-green"].$type;
      tokens.font.feather.$description = "";
    }),
  ({ output }) => {
    assert.match(output, /color\.eager-green must use type color/u);
    assert.match(output, /font\.feather must have a description/u);
  },
);

withFixture(
  (root) =>
    mutateTokens(root, (tokens) => {
      tokens.$extensions["com.vocora.design-system"].buttonRadius = "8px";
      delete tokens.themes.dark.action.error;
    }),
  ({ output }) => {
    assert.match(output, /The canonical button radius must be 13px/u);
    assert.match(output, /Theme key parity failed for group 'action'/u);
  },
);

withFixture(
  (root) => {
    const variablesPath = path.join(root, "references", "variables.scss");
    fs.writeFileSync(
      variablesPath,
      fs
        .readFileSync(variablesPath, "utf8")
        .replace(
          "--color-spark-blue: #1cb0f6;",
          "--color-spark-blue: #000000;",
        ),
    );
  },
  ({ output }) =>
    assert.match(
      output,
      /variables\.scss must map --color-spark-blue to #1CB0F6/u,
    ),
);

withFixture(
  (root) => {
    const designPath = path.join(root, "references", "DESIGN.md");
    fs.writeFileSync(
      designPath,
      fs
        .readFileSync(designPath, "utf8")
        .replace(
          "Reusable product-specific primitive?",
          "Reusable product primitive?",
        ),
    );
  },
  ({ output }) =>
    assert.match(
      output,
      /DESIGN\.md is missing decision-tree step: Reusable product-specific primitive\?/u,
    ),
);

withFixture(
  (root) => {
    const designPath = path.join(root, "references", "DESIGN.md");
    fs.writeFileSync(
      designPath,
      fs
        .readFileSync(designPath, "utf8")
        .replaceAll("voco-navigation-button", "removed-navigation-button"),
    );
  },
  ({ status, output }) => {
    assert.notEqual(status, 0);
    assert.match(
      output,
      /DESIGN\.md is missing button selector: voco-navigation-button/u,
    );
  },
);

withFixture(
  (root) => {
    const designPath = path.join(root, "references", "DESIGN.md");
    fs.writeFileSync(
      designPath,
      fs
        .readFileSync(designPath, "utf8")
        .replace("`(activated)` is the public action event", "The public action event"),
    );
  },
  ({ status, output }) => {
    assert.notEqual(status, 0);
    assert.match(
      output,
      /DESIGN\.md must document the public voco activated event/u,
    );
  },
);

withFixture(
  (root) => {
    const designPath = path.join(root, "references", "DESIGN.md");
    fs.writeFileSync(
      designPath,
      fs
        .readFileSync(designPath, "utf8")
        .replaceAll("VocoWarningButtonComponent", "RemovedWarningButton"),
    );
  },
  ({ status, output }) => {
    assert.notEqual(status, 0);
    assert.match(
      output,
      /DESIGN\.md is missing public button class: VocoWarningButtonComponent/u,
    );
  },
);

for (const [contractText, expected] of [
  [
    "removes both native `href` and `routerLink`",
    /DESIGN\.md must define disabled Voco link behavior/u,
  ],
  [
    "`voco-audio-button` is the square icon-oriented audio transport control",
    /DESIGN\.md must distinguish square audio controls from textual audio actions/u,
  ],
  [
    "not a generic Voco\nbutton replacement",
    /DESIGN\.md must keep vocoButtonInteraction selection-only/u,
  ],
]) {
  withFixture(
    (root) => {
      const designPath = path.join(root, "references", "DESIGN.md");
      fs.writeFileSync(
        designPath,
        fs.readFileSync(designPath, "utf8").replace(contractText, "removed contract"),
      );
    },
    ({ status, output }) => {
      assert.notEqual(status, 0);
      assert.match(output, expected);
    },
  );
}

console.log("K2 design-system validator tests passed.");
